### PR TITLE
Working on #110

### PR DESCRIPTION
**Pull request by mindstorm38**
Working on #110

Poke @JamiKettunen

### CODE REVIEW

**Comment by mindstorm38** [comment]
This first iteration ad a warning for non-glibc systems.

**Comment by JamiKettunen** [comment]
Works as normal on `glibc` still and gives the following on `musl` with `portablemc start` (as expected):
```
$ portablemc start                                                                                                                                                                             19
[  OK  ] Resolved version 1.19.2.    
[  OK  ] Loaded version JAR.    
[  OK  ] Checked 3390 assets.
[  OK  ] Loaded pretty logger.
[  OK  ] Loaded 54 libraries. 
[FAILED] No JVM download was found for your libc, use --jvm argument to manually set the path to your JVM executable.
```

I'm just wondering if we could try to use `java` executable from `$PATH` as a  default if it exists as otherwise you'd have to provide `--jvm` yourself anyway; this `java` symlink can be managed on Void Linux with e.g. this:
```
# xbps-alternatives -s openjdk17-jre java
openjdk17-jre-17.0.5+7_1: applying 'java' alternatives group
Creating 'java' alternatives group symlink: /usr/bin/java -> /usr/lib/jvm/openjdk17/bin/java
Creating 'java' alternatives group symlink: /usr/bin/jfr -> /usr/lib/jvm/openjdk17/bin/jfr
Creating 'java' alternatives group symlink: /usr/bin/keytool -> /usr/lib/jvm/openjdk17/bin/keytool
Creating 'java' alternatives group symlink: /usr/bin/rmiregistry -> /usr/lib/jvm/openjdk17/bin/rmiregistry
```

**Comment by JamiKettunen** [comment]
Not having much luck under `aarch64-musl` so far with getting it to run with the following commands with the changes here:
```
rm -r /tmp/lwjgl$USER
portablemc start \
	--jvm java \
	--lwjgl 3.2.3 \
	--exclude-lib lwjgl-glfw::natives-linux --exclude-lib lwjgl-openal::natives-linux --exclude-lib lwjgl-jemalloc::natives-linux \
	--include-bin /usr/lib/libglfw.so.3.3 --include-bin /usr/lib/libopenal.so.1.22.2 --include-bin /usr/lib/libjemalloc.so.2 \
	1.16.5
```
`/tmp/lwjglsomainline/3.2.3-build-13/libglfw.so` is still getting extracted and links to `glibc` so it won't load under `musl`.

**Comment by mindstorm38** [comment]
Ahhh ok maybe "natives-linux" is not right under aarch64 when using custom lwjgl, it's "natives-linux-arm64" according  to:
 https://github.com/mindstorm38/portablemc/blob/master/src/core/portablemc/cli.py#L661

**Comment by JamiKettunen** [comment]
I'll debug this later today, nothing changed after adding versioning like `3.2.3` or `3.2.3-build-13` to `--exclude-lib` (in between empty `::`) or using `natives-linux-arm64` instead of `natives-linux`.

Would be great to have more user input checking for errors (in case items specified aren't in lists and won't change anything for runtime etc.) and perhaps an option to list all the `--exclude-lib` options without having to look at the script

**Comment by mindstorm38** [comment]
The input is already validated, but being a little more verbose would not be to much, I'll add this later today.

**Comment by mindstorm38** [comment]
I just added informative messages for every pattern (unused and usage count if so).

**Comment by JamiKettunen** [comment]
So with `--exclude-lib lwjgl-glfw:3.2.3` as well as `--exclude-lib lwjgl-glfw`: `Library exclusion 'lwjgl-glfw:' matched 4 libraries.`, but adding debugging in the `LibrarySpecifierFilter` `matches()` method shows that the `spec.classifier` is always `None`, thus a `self.classifier` of `natives-linux` or whatever never matched anything.

**Comment by mindstorm38** [comment]
`--exclude-lib lwjgl-glfw::natives-linux-arm64` should set `self.classifier = "natives-linux-arm64"`, at least it work for me on windows so I don't understand why it doesn't work for you

**Comment by JamiKettunen** [comment]
`self.classifier` is set just fine, it's the **spec ones** that are all `None` in my case

**Comment by mindstorm38** [comment]
Ok I see, I suspect the lwjgl argument. Could you get the dump of the metadata file (should be next to the standard file in version directory)?

**Comment by mindstorm38** [comment]
Should work now!

**Comment by JamiKettunen** [comment]
Indeed with `--exclude-lib lwjgl-glfw::natives-linux` it now shows `Library exclusion 'lwjgl-glfw::natives-linux' matched 1 libraries.`, yet similar to before `/tmp/lwjgl$USER/3.2.3-build-13/libglfw.so` is still being extracted upon `portablemc start` after removing `/tmp/lwjgl$USER` or even `~/.minecraft` to get a fully clean setup.

Btw I just tested this on `x86_64-musl` and there it works properly right now, so it's an `aarch64-musl` exclusive problem it seems, **but**  I think as part of the cleanup operations `/tmp/lwjgl$USER` should be removed as it's a part of the state of the game and not cleaned up afterwards, or else it'll try to use the already extracted libraries for e.g. `libglfw.so` next time even when `--exclude-lib lwjgl-glfw::natives-linux` is being specified; perhaps this should only be done if libraries are being excluded?

**Comment by mindstorm38** [comment]
I think it's now a problem with the classifier you're excluding, because on arm the classifiers are both `natives-linux-arm64` and `natives-linux-arm32`, not `natives-linux`. Perhaps I should support globbing like `natives-linux*`?

**Comment by mindstorm38** [comment]
> **but** I think as part of the cleanup operations `/tmp/lwjgl$USER` should be removed as it's a part of the state of the game and not cleaned up afterwards

The issue is that this directory is managed by the LWJGL runtime, so I can't know the path from Python. Maybe it's a LWJGL problem...

**Comment by JamiKettunen** [comment]
> I think it's now a problem with the classifier you're excluding, because on arm the classifiers are both `natives-linux-arm64` and `natives-linux-arm32`, not `natives-linux`.

Ah you're correct, `--exclude-lib lwjgl-glfw::natives-linux-arm64` did the trick!

> Perhaps I should support globbing like `natives-linux*`?

That would be great, less architecture-specific edge cases like this to worry about from a user's perspective; I wonder if it should even glob the end always? Then you could even forget about the platform specific part and just specify `--exclude-lib lwjgl-glfw::natives` and it'll work on windows, linux etc. across all CPU architectures.

In my case specifying both `--exclude-lib lwjgl-glfw::natives-linux --exclude-lib lwjgl-glfw::natives-linux-arm64` didn't cause any harm.

**Comment by mindstorm38** [comment]
> That would be great, less architecture-specific edge cases like this to worry about from a user's perspective; I wonder if it should even glob the end always? Then you could even forget about the platform specific part and just specify `--exclude-lib lwjgl-glfw::natives` and it'll work on windows, linux etc. across all CPU architectures.

Good idea

File: src/core/portablemc/cli.py [hunk 1]
```diff
@@ -499,10 +499,19 @@ def _predicate(spec: "LibrarySpecifier") -> bool:
...
-            version.prepare_jvm()
-            print_task("OK", "start.jvm.loaded", {"version": version.jvm_version}, done=True)
+            try:
+                print_task("", "start.jvm.loading")
+                version.prepare_jvm()
+                print_task("OK", "start.jvm.loaded", {"version": version.jvm_version}, done=True)
+            except JvmLoadingError:
+                ns.jvm = shutil.which(get_jvm_bin_filename())
+                if ns.jvm is not None:
+                    print_task("OK", "start.jvm.system_fallback", {"path": ns.jvm})
```
Comment (JamiKettunen) [inline src/core/portablemc/cli.py:509]: Needs `, done=True` to avoid getting overwritten by next `print_task()` call I believe?

Comment (mindstorm38) [reply src/core/portablemc/cli.py:509]: right!

**Comment by JamiKettunen** [comment]
Btw I was gonna check out the [`forcegl20` fabric mod](https://modrinth.com/mod/forcegl20) but with the changes here it appears the `fabric` add-on (v2.0.1) has stopped working (`portablemc start fabric:`):
```
[      ] Loading libraries... Traceback (most recent call last):
  File "/usr/bin/portablemc", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3.10/site-packages/portablemc/cli.py", line 145, in main
    cmd(handler, ns)
  File "/usr/lib/python3.10/site-packages/portablemc/cli.py", line 350, in cmd
    handler(ns, new_context(ns))
  File "/usr/lib/python3.10/site-packages/portablemc/cli.py", line 797, in wrapper
    return func(orig_func, *w_args, **w_kwargs)
  File "/usr/lib/python3.10/site-packages/portablemc_fabric/__init__.py", line 29, in cmd_start
    return old(ns, ctx)
  File "/usr/lib/python3.10/site-packages/portablemc/cli.py", line 485, in cmd_start
    version.prepare_libraries(predicate=libraries_predicate)
  File "/usr/lib/python3.10/site-packages/portablemc/__init__.py", line 463, in prepare_libraries
    lib_path_raw = lib_spec.jar_file_path()
  File "/usr/lib/python3.10/site-packages/portablemc/__init__.py", line 1636, in jar_file_path
    return path.join((*self.group.split("."), self.artifact, self.version, file_name))
  File "/usr/lib/python3.10/posixpath.py", line 76, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not tuple
```

**Comment by JamiKettunen** [comment]
> Should work! (I must add anti-regression tests for this)

+1, code coverage checking in general would be great, the more the better chance to catch random errors like the above that might've been introduced.

Also great news from my side, I can run even fully vanilla 1.19.2 with `MESA_GL_VERSION_OVERRIDE=3.2 MESA_GLSL_VERSION_OVERRIDE=150` in env! Thus my "final" command currently is:
```sh
MESA_GL_VERSION_OVERRIDE=3.2 MESA_GLSL_VERSION_OVERRIDE=150 portablemc start --lwjgl 3.2.3 \
    --exclude-lib lwjgl-glfw::natives-linux-arm64 --include-bin /lib/libglfw.so.3 \
    --exclude-lib lwjgl-openal::natives-linux-arm64 --include-bin /lib/libopenal.so.1
```
Excluding `jemalloc` didn't seem to be necessary in the end, the game launches and I can go in-game and don't see anything different in the logs with or without `--exclude-lib lwjgl-jemalloc::natives-linux-arm64 --include-bin /lib/libjemalloc.so.2` so :shrug:

**Comment by mindstorm38** [comment]
Playing with coverage... Tell me if you know how to improve :) I decided to merge it with this PR because I didn't want to wait before having coverage on tests made for it

**Comment by mindstorm38** [comment]
> That would be great, less architecture-specific edge cases like this to worry about from a user's perspective; I wonder if it should even glob the end always? Then you could even forget about the platform specific part and just specify `--exclude-lib lwjgl-glfw::natives` and it'll work on windows, linux etc. across all CPU architectures.

Implemented! (to test)

This PR seems to be on track for merging... If I don't miss anything, maybe start a short code review period?

**Comment by mindstorm38** [comment]
Note that the codecov report is not working here because it was just introduced in this PR, this should work for upcoming PRs

**Comment by JamiKettunen** [comment]
> Implemented! (to test)

Seems to work just fine with the previous command or `--exclude-lib lwjgl-glfw::natives --exclude-lib lwjgl-openal::natives`:
```
[  OK  ] Loaded 73 libraries. 
         Library exclusion 'lwjgl-glfw::natives' matched 3 libraries.
         Library exclusion 'lwjgl-openal::natives' matched 3 libraries.
```

> This PR seems to be on track for merging... If I don't miss anything, maybe start a short code review period?

Sounds like a plan! `--exclude-lib` help should perhaps include an example such as `lwjgl-glfw::natives`

File: src/core/portablemc/__init__.py [hunk 2]
```diff
@@ -775,6 +790,19 @@ def cleanup():
...
+        for bin_file in self.bin_files:
+            bin_name = path.basename(bin_file)
+            # Here we try to remove the version numbers of .so files.
+            so_idx = bin_name.rfind(".so")
+            if so_idx >= 0:
+                bin_name = bin_name[:so_idx + len(".so")]
+            # Try to symlink the file in the bin dir, and fallback to simple copy.
+            bin_dst_file = path.join(bin_dir, bin_name)
+            try:
+                os.symlink(bin_file, bin_dst_file)
```
Comment (JamiKettunen) [inline src/core/portablemc/__init__.py:802]: Before even this is attempted it should be ensured that `bin_file` even exists on the filesystem and perhaps throw an error otherwise?

File: src/core/portablemc/__init__.py [hunk 3]
```diff
@@ -775,6 +790,22 @@ def cleanup():
                             with open(path.join(bin_dir, native_name), "wb") as native_file:
                                 shutil.copyfileobj(native_zip_file, native_file)
 
+        for bin_file in self.bin_files:
+            # Resolve a potential symlink and check if the binary exists before linking.
+            bin_file = path.realpath(bin_file)
+            if path.isfile(bin_file):
```
Comment (JamiKettunen) [inline src/core/portablemc/__init__.py:796]: Hm not sure how I feel about this, you can still give `--include-bin` a file that doesn't exist and it'll be silently ignored with no warning message or anything.

Comment (mindstorm38) [reply src/core/portablemc/__init__.py:796]: You're right, I didn't know how to properly implement a feedback... Maybe a validation before starting the game, but I prefere an API validation. Maybe raising an error?

Comment (JamiKettunen) [reply src/core/portablemc/__init__.py:796]: Raising an error would make sense in my eyes [and is what I suggested before](https://github.com/mindstorm38/portablemc/pull/112#discussion_r1008039070) :)

Comment (mindstorm38) [reply src/core/portablemc/__init__.py:796]: I missed it!

File: src/core/portablemc/__init__.py [hunk 4]
```diff
@@ -1611,6 +1613,19 @@ class JvmLoadingError(BaseError):
     UNSUPPORTED_LIBC = "unsupported_libc"
 
 
+class BinaryNotFound(Exception):
+
+    """
+    This type of error can be raised by `Start.start(...)` to signal the caller that
+    on of it's additional binary doesn't exists and therefore can't be
```
Comment (JamiKettunen) [inline src/core/portablemc/__init__.py:1620]: `one of the specified additional binaries` or something similar?

File: src/core/portablemc/cli.py [hunk 5]
```diff
@@ -568,6 +568,9 @@ def _predicate(spec: "LibrarySpecifier") -> bool:
     except JvmLoadingError as err:
         print_task("FAILED", f"start.jvm.error.{err.code}", done=True)
         sys.exit(EXIT_JVM_LOADING_ERROR)
+    except BinaryNotFound as err:
+        print_task("FAILED", f"start.additional_binary_not_found", {"bin": err.bin_file}, done=True)
```
Comment (JamiKettunen) [inline src/core/portablemc/cli.py:572]: Very minor but `f-string` not needed

**Comment by mindstorm38** [comment]
Btw I'm thinking of renumbering the next version to 3.2 because these changes are no longer patch

**Review by mindstorm38 — COMMENTED** [review]
I reviewed again my changes, and I think we're ready to merge!
